app/models/composition_saver.rb

```diff
@@ -0,0 +1,150 @@
+class CompositionSaver
+  attr_reader :error_type, :error_value, :composition
+
+  class Error < StandardError; end
+
+  def initialize(user:, session_id:)
+    @user = user
+    @session_id = session_id
+  end
+
+  def save(data)
+    if data[:player_id] || data[:player_name]
+      player = init_player(data)
+      unless player.persisted? && !player.changed? || player.save
+        @error_type = 'player'
+        @error_value = player.errors
+        return
+      end
+    end
+
+    map = if data[:map_segment_id]
+      map_segment = MapSegment.find(data[:map_segment_id])
+      map_segment.map
+    elsif data[:map_id]
+      Map.find(data[:map_id])
+    end
+
+    if data[:composition_id] || map
+      @composition = init_composition(data, map: map)
+      unless @composition.persisted? || @composition.save
+        @error_type = 'composition'
+        @error_value = @composition.errors
+        return
+      end
+    end
+
+    if player && @composition
+      comp_player = init_composition_player(data, player: player,
+                                            composition: @composition)
+      unless comp_player.persisted? || comp_player.save
+        @error_type = 'composition_player'
+        @error_value = comp_player.errors
+        return
+      end
+    end
+
+    if player && map_segment && data[:hero_id]
+      selection = init_player_selection(data, composition: @composition,
+                                        player: player, map_segment: map_segment)
+      unless selection.persisted? && !selection.changed? || selection.save
+        @error_type = 'player_selection'
+        @error_value = selection.errors
+        return
+      end
+    end
+
+    true
+  end
```

Comment:
> I feel like this is pretty complex to read.

I agree! I feel like this is an improvement over the logic being in the controller, but I think we can continue to improve this model. 

> so they always have a select list of their previously created players

I agree! The idea had come up before of autocompleting player names based on past compositions, or when you start a new composition you get text fields already filled with your past player names. Just moving to select menus sounds easier, though. I'll file an issue!